test/http2_test/test_data_frame_padding.py

```diff
@@ -0,0 +1,113 @@
+    self._base_server._handlers['WindowUpdated'] = self.on_window_update
+    self._base_server._handlers['RequestReceived'] = self.on_request_received
+    self._base_server._handlers['SendDone'] = self.on_send_done
+    self._base_server._handlers['ConnectionMade'] = self.on_connection_made
+    self._base_server._handlers['ConnectionLost'] = self.on_connection_lost
+    self._serving_active_connection = True
+    self._reset_connection_state()
+    if use_padding:
+      self._pad_length = 255
+    else:
+      self._pad_length = None
+
+  def get_base_server(self):
+    return self._base_server
+
+  def _reset_connection_state(self):
+    # maps stream ids to total flow control updates received
+    self._total_updates = {}
+    self._total_updates[0] = 0
+    self._read_chunk_size = 5
+    # currently active streams with connections sent on them
+    self._cur_sending = []
+
+  def on_connection_made(self):
+    self._base_server.on_connection_made_default()
+    self._reset_connection_state()
+    assert not self._serving_active_connection
+    self._serving_active_connection = True
+
+  def on_connection_lost(self, reason):
+    self._base_server.on_connection_lost(reason)
+    self._serving_active_connection = False
+
+  def on_data_received(self, event):
+    logging.info('on data received. Stream id: %d. Data length: %d' % (event.stream_id, len(event.data)))
+    self._base_server.on_data_received_default(event)
+    if len(event.data) == 0:
+      return
+    sr = self._base_server.parse_received_data(event.stream_id)
+    stream_bytes = ''
+    # Check if full grpc msg has been read into the recv buffer yet
+    if sr:
+      response_data = self._base_server.default_response_data(sr.response_size)
+      logging.info('Stream id: %d. total resp size: %d' % (event.stream_id, len(response_data)))
+      assert event.stream_id not in self._cur_sending
+      self._cur_sending.append(event.stream_id)
+      self._ready_to_send = True
+      self._base_server.setup_send(response_data , event.stream_id, pad_length=self._pad_length, read_chunk_size=self._read_chunk_size)
+
+  def on_send_done(self, stream_id):
+    assert stream_id in self._cur_sending
+    self._cur_sending.remove(stream_id)
+    self._base_server.on_send_done_default(stream_id)
+
+  def on_request_received(self, event):
+    self._base_server.on_request_received_default(event)
+    logging.info('on request received. Stream id: %s.' % event.stream_id)
+    self._total_updates[event.stream_id] = 0
+
+  def on_window_update(self, event):
+    logging.info('on window update. Stream id: %s. Delta: %s' % (event.stream_id, event.delta))
+    self._total_updates[event.stream_id] += event.delta
+    total = self._total_updates[event.stream_id]
+    logging.info('... - total updates for stream %d : %d' % (event.stream_id, total))
+    p = []
+    if event.stream_id == 0:
+      p = self._cur_sending
+    else:
+      p = [event.stream_id]
+    for p in self._cur_sending:
```

Comment:
woops, fixed, this is a mistake.

This tries to resume sending only on the stream which the window update is for, or all if the window update is a connection update - same outcome but mean to limit resuming of sending to only refreshed streams.